<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN"
    "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en">
<head>
<meta http-equiv="Content-Type" content="application/xhtml+xml; charset=UTF-8" />
<meta name="generator" content="AsciiDoc 8.6.8" />
<title>RNA-Seq De novo Assembly Using Trinity</title>
<style type="text/css">
/* Shared CSS for AsciiDoc xhtml11 and html5 backends */

/* Default font. */
body {
  font-family: Georgia,serif;
}

/* Title font. */
h1, h2, h3, h4, h5, h6,
div.title, caption.title,
thead, p.table.header,
#toctitle,
#author, #revnumber, #revdate, #revremark,
#footer {
  font-family: Arial,Helvetica,sans-serif;
}

body {
  margin: 1em 5% 1em 5%;
}

a {
  color: blue;
  text-decoration: underline;
}
a:visited {
  color: fuchsia;
}

em {
  font-style: italic;
  color: navy;
}

strong {
  font-weight: bold;
  color: #083194;
}

h1, h2, h3, h4, h5, h6 {
  color: #527bbd;
  margin-top: 1.2em;
  margin-bottom: 0.5em;
  line-height: 1.3;
}

h1, h2, h3 {
  border-bottom: 2px solid silver;
}
h2 {
  padding-top: 0.5em;
}
h3 {
  float: left;
}
h3 + * {
  clear: left;
}
h5 {
  font-size: 1.0em;
}

div.sectionbody {
  margin-left: 0;
}

hr {
  border: 1px solid silver;
}

p {
  margin-top: 0.5em;
  margin-bottom: 0.5em;
}

ul, ol, li > p {
  margin-top: 0;
}
ul > li     { color: #aaa; }
ul > li > * { color: black; }

.monospaced, code, pre {
  font-family: "Courier New", Courier, monospace;
  font-size: inherit;
  color: navy;
  padding: 0;
  margin: 0;
}


#author {
  color: #527bbd;
  font-weight: bold;
  font-size: 1.1em;
}
#email {
}
#revnumber, #revdate, #revremark {
}

#footer {
  font-size: small;
  border-top: 2px solid silver;
  padding-top: 0.5em;
  margin-top: 4.0em;
}
#footer-text {
  float: left;
  padding-bottom: 0.5em;
}
#footer-badges {
  float: right;
  padding-bottom: 0.5em;
}

#preamble {
  margin-top: 1.5em;
  margin-bottom: 1.5em;
}
div.imageblock, div.exampleblock, div.verseblock,
div.quoteblock, div.literalblock, div.listingblock, div.sidebarblock,
div.admonitionblock {
  margin-top: 1.0em;
  margin-bottom: 1.5em;
}
div.admonitionblock {
  margin-top: 2.0em;
  margin-bottom: 2.0em;
  margin-right: 10%;
  color: #606060;
}

div.content { /* Block element content. */
  padding: 0;
}

/* Block element titles. */
div.title, caption.title {
  color: #527bbd;
  font-weight: bold;
  text-align: left;
  margin-top: 1.0em;
  margin-bottom: 0.5em;
}
div.title + * {
  margin-top: 0;
}

td div.title:first-child {
  margin-top: 0.0em;
}
div.content div.title:first-child {
  margin-top: 0.0em;
}
div.content + div.title {
  margin-top: 0.0em;
}

div.sidebarblock > div.content {
  background: #ffffee;
  border: 1px solid #dddddd;
  border-left: 4px solid #f0f0f0;
  padding: 0.5em;
}

div.listingblock > div.content {
  border: 1px solid #dddddd;
  border-left: 5px solid #f0f0f0;
  background: #f8f8f8;
  padding: 0.5em;
}

div.quoteblock, div.verseblock {
  padding-left: 1.0em;
  margin-left: 1.0em;
  margin-right: 10%;
  border-left: 5px solid #f0f0f0;
  color: #888;
}

div.quoteblock > div.attribution {
  padding-top: 0.5em;
  text-align: right;
}

div.verseblock > pre.content {
  font-family: inherit;
  font-size: inherit;
}
div.verseblock > div.attribution {
  padding-top: 0.75em;
  text-align: left;
}
/* DEPRECATED: Pre version 8.2.7 verse style literal block. */
div.verseblock + div.attribution {
  text-align: left;
}

div.admonitionblock .icon {
  vertical-align: top;
  font-size: 1.1em;
  font-weight: bold;
  text-decoration: underline;
  color: #527bbd;
  padding-right: 0.5em;
}
div.admonitionblock td.content {
  padding-left: 0.5em;
  border-left: 3px solid #dddddd;
}

div.exampleblock > div.content {
  border-left: 3px solid #dddddd;
  padding-left: 0.5em;
}

div.imageblock div.content { padding-left: 0; }
span.image img { border-style: none; }
a.image:visited { color: white; }

dl {
  margin-top: 0.8em;
  margin-bottom: 0.8em;
}
dt {
  margin-top: 0.5em;
  margin-bottom: 0;
  font-style: normal;
  color: navy;
}
dd > *:first-child {
  margin-top: 0.1em;
}

ul, ol {
    list-style-position: outside;
}
ol.arabic {
  list-style-type: decimal;
}
ol.loweralpha {
  list-style-type: lower-alpha;
}
ol.upperalpha {
  list-style-type: upper-alpha;
}
ol.lowerroman {
  list-style-type: lower-roman;
}
ol.upperroman {
  list-style-type: upper-roman;
}

div.compact ul, div.compact ol,
div.compact p, div.compact p,
div.compact div, div.compact div {
  margin-top: 0.1em;
  margin-bottom: 0.1em;
}

tfoot {
  font-weight: bold;
}
td > div.verse {
  white-space: pre;
}

div.hdlist {
  margin-top: 0.8em;
  margin-bottom: 0.8em;
}
div.hdlist tr {
  padding-bottom: 15px;
}
dt.hdlist1.strong, td.hdlist1.strong {
  font-weight: bold;
}
td.hdlist1 {
  vertical-align: top;
  font-style: normal;
  padding-right: 0.8em;
  color: navy;
}
td.hdlist2 {
  vertical-align: top;
}
div.hdlist.compact tr {
  margin: 0;
  padding-bottom: 0;
}

.comment {
  background: yellow;
}

.footnote, .footnoteref {
  font-size: 0.8em;
}

span.footnote, span.footnoteref {
  vertical-align: super;
}

#footnotes {
  margin: 20px 0 20px 0;
  padding: 7px 0 0 0;
}

#footnotes div.footnote {
  margin: 0 0 5px 0;
}

#footnotes hr {
  border: none;
  border-top: 1px solid silver;
  height: 1px;
  text-align: left;
  margin-left: 0;
  width: 20%;
  min-width: 100px;
}

div.colist td {
  padding-right: 0.5em;
  padding-bottom: 0.3em;
  vertical-align: top;
}
div.colist td img {
  margin-top: 0.3em;
}

@media print {
  #footer-badges { display: none; }
}

#toc {
  margin-bottom: 2.5em;
}

#toctitle {
  color: #527bbd;
  font-size: 1.1em;
  font-weight: bold;
  margin-top: 1.0em;
  margin-bottom: 0.1em;
}

div.toclevel0, div.toclevel1, div.toclevel2, div.toclevel3, div.toclevel4 {
  margin-top: 0;
  margin-bottom: 0;
}
div.toclevel2 {
  margin-left: 2em;
  font-size: 0.9em;
}
div.toclevel3 {
  margin-left: 4em;
  font-size: 0.9em;
}
div.toclevel4 {
  margin-left: 6em;
  font-size: 0.9em;
}

span.aqua { color: aqua; }
span.black { color: black; }
span.blue { color: blue; }
span.fuchsia { color: fuchsia; }
span.gray { color: gray; }
span.green { color: green; }
span.lime { color: lime; }
span.maroon { color: maroon; }
span.navy { color: navy; }
span.olive { color: olive; }
span.purple { color: purple; }
span.red { color: red; }
span.silver { color: silver; }
span.teal { color: teal; }
span.white { color: white; }
span.yellow { color: yellow; }

span.aqua-background { background: aqua; }
span.black-background { background: black; }
span.blue-background { background: blue; }
span.fuchsia-background { background: fuchsia; }
span.gray-background { background: gray; }
span.green-background { background: green; }
span.lime-background { background: lime; }
span.maroon-background { background: maroon; }
span.navy-background { background: navy; }
span.olive-background { background: olive; }
span.purple-background { background: purple; }
span.red-background { background: red; }
span.silver-background { background: silver; }
span.teal-background { background: teal; }
span.white-background { background: white; }
span.yellow-background { background: yellow; }

span.big { font-size: 2em; }
span.small { font-size: 0.6em; }

span.underline { text-decoration: underline; }
span.overline { text-decoration: overline; }
span.line-through { text-decoration: line-through; }

div.unbreakable { page-break-inside: avoid; }


/*
 * xhtml11 specific
 *
 * */

div.tableblock {
  margin-top: 1.0em;
  margin-bottom: 1.5em;
}
div.tableblock > table {
  border: 3px solid #527bbd;
}
thead, p.table.header {
  font-weight: bold;
  color: #527bbd;
}
p.table {
  margin-top: 0;
}
/* Because the table frame attribute is overriden by CSS in most browsers. */
div.tableblock > table[frame="void"] {
  border-style: none;
}
div.tableblock > table[frame="hsides"] {
  border-left-style: none;
  border-right-style: none;
}
div.tableblock > table[frame="vsides"] {
  border-top-style: none;
  border-bottom-style: none;
}


/*
 * html5 specific
 *
 * */

table.tableblock {
  margin-top: 1.0em;
  margin-bottom: 1.5em;
}
thead, p.tableblock.header {
  font-weight: bold;
  color: #527bbd;
}
p.tableblock {
  margin-top: 0;
}
table.tableblock {
  border-width: 3px;
  border-spacing: 0px;
  border-style: solid;
  border-color: #527bbd;
  border-collapse: collapse;
}
th.tableblock, td.tableblock {
  border-width: 1px;
  padding: 4px;
  border-style: solid;
  border-color: #527bbd;
}

table.tableblock.frame-topbot {
  border-left-style: hidden;
  border-right-style: hidden;
}
table.tableblock.frame-sides {
  border-top-style: hidden;
  border-bottom-style: hidden;
}
table.tableblock.frame-none {
  border-style: hidden;
}

th.tableblock.halign-left, td.tableblock.halign-left {
  text-align: left;
}
th.tableblock.halign-center, td.tableblock.halign-center {
  text-align: center;
}
th.tableblock.halign-right, td.tableblock.halign-right {
  text-align: right;
}

th.tableblock.valign-top, td.tableblock.valign-top {
  vertical-align: top;
}
th.tableblock.valign-middle, td.tableblock.valign-middle {
  vertical-align: middle;
}
th.tableblock.valign-bottom, td.tableblock.valign-bottom {
  vertical-align: bottom;
}


/*
 * manpage specific
 *
 * */

body.manpage h1 {
  padding-top: 0.5em;
  padding-bottom: 0.5em;
  border-top: 2px solid silver;
  border-bottom: 2px solid silver;
}
body.manpage h2 {
  border-style: none;
}
body.manpage div.sectionbody {
  margin-left: 3em;
}

@media print {
  body.manpage div#toc { display: none; }
}


</style>
<script type="text/javascript">
/*<![CDATA[*/
var asciidoc = {  // Namespace.

/////////////////////////////////////////////////////////////////////
// Table Of Contents generator
/////////////////////////////////////////////////////////////////////

/* Author: Mihai Bazon, September 2002
 * http://students.infoiasi.ro/~mishoo
 *
 * Table Of Content generator
 * Version: 0.4
 *
 * Feel free to use this script under the terms of the GNU General Public
 * License, as long as you do not remove or alter this notice.
 */

 /* modified by Troy D. Hanson, September 2006. License: GPL */
 /* modified by Stuart Rackham, 2006, 2009. License: GPL */

// toclevels = 1..4.
toc: function (toclevels) {

  function getText(el) {
    var text = "";
    for (var i = el.firstChild; i != null; i = i.nextSibling) {
      if (i.nodeType == 3 /* Node.TEXT_NODE */) // IE doesn't speak constants.
        text += i.data;
      else if (i.firstChild != null)
        text += getText(i);
    }
    return text;
  }

  function TocEntry(el, text, toclevel) {
    this.element = el;
    this.text = text;
    this.toclevel = toclevel;
  }

  function tocEntries(el, toclevels) {
    var result = new Array;
    var re = new RegExp('[hH]([1-'+(toclevels+1)+'])');
    // Function that scans the DOM tree for header elements (the DOM2
    // nodeIterator API would be a better technique but not supported by all
    // browsers).
    var iterate = function (el) {
      for (var i = el.firstChild; i != null; i = i.nextSibling) {
        if (i.nodeType == 1 /* Node.ELEMENT_NODE */) {
          var mo = re.exec(i.tagName);
          if (mo && (i.getAttribute("class") || i.getAttribute("className")) != "float") {
            result[result.length] = new TocEntry(i, getText(i), mo[1]-1);
          }
          iterate(i);
        }
      }
    }
    iterate(el);
    return result;
  }

  var toc = document.getElementById("toc");
  if (!toc) {
    return;
  }

  // Delete existing TOC entries in case we're reloading the TOC.
  var tocEntriesToRemove = [];
  var i;
  for (i = 0; i < toc.childNodes.length; i++) {
    var entry = toc.childNodes[i];
    if (entry.nodeName.toLowerCase() == 'div'
     && entry.getAttribute("class")
     && entry.getAttribute("class").match(/^toclevel/))
      tocEntriesToRemove.push(entry);
  }
  for (i = 0; i < tocEntriesToRemove.length; i++) {
    toc.removeChild(tocEntriesToRemove[i]);
  }

  // Rebuild TOC entries.
  var entries = tocEntries(document.getElementById("content"), toclevels);
  for (var i = 0; i < entries.length; ++i) {
    var entry = entries[i];
    if (entry.element.id == "")
      entry.element.id = "_toc_" + i;
    var a = document.createElement("a");
    a.href = "#" + entry.element.id;
    a.appendChild(document.createTextNode(entry.text));
    var div = document.createElement("div");
    div.appendChild(a);
    div.className = "toclevel" + entry.toclevel;
    toc.appendChild(div);
  }
  if (entries.length == 0)
    toc.parentNode.removeChild(toc);
},


/////////////////////////////////////////////////////////////////////
// Footnotes generator
/////////////////////////////////////////////////////////////////////

/* Based on footnote generation code from:
 * http://www.brandspankingnew.net/archive/2005/07/format_footnote.html
 */

footnotes: function () {
  // Delete existing footnote entries in case we're reloading the footnodes.
  var i;
  var noteholder = document.getElementById("footnotes");
  if (!noteholder) {
    return;
  }
  var entriesToRemove = [];
  for (i = 0; i < noteholder.childNodes.length; i++) {
    var entry = noteholder.childNodes[i];
    if (entry.nodeName.toLowerCase() == 'div' && entry.getAttribute("class") == "footnote")
      entriesToRemove.push(entry);
  }
  for (i = 0; i < entriesToRemove.length; i++) {
    noteholder.removeChild(entriesToRemove[i]);
  }

  // Rebuild footnote entries.
  var cont = document.getElementById("content");
  var spans = cont.getElementsByTagName("span");
  var refs = {};
  var n = 0;
  for (i=0; i<spans.length; i++) {
    if (spans[i].className == "footnote") {
      n++;
      var note = spans[i].getAttribute("data-note");
      if (!note) {
        // Use [\s\S] in place of . so multi-line matches work.
        // Because JavaScript has no s (dotall) regex flag.
        note = spans[i].innerHTML.match(/\s*\[([\s\S]*)]\s*/)[1];
        spans[i].innerHTML =
          "[<a id='_footnoteref_" + n + "' href='#_footnote_" + n +
          "' title='View footnote' class='footnote'>" + n + "</a>]";
        spans[i].setAttribute("data-note", note);
      }
      noteholder.innerHTML +=
        "<div class='footnote' id='_footnote_" + n + "'>" +
        "<a href='#_footnoteref_" + n + "' title='Return to text'>" +
        n + "</a>. " + note + "</div>";
      var id =spans[i].getAttribute("id");
      if (id != null) refs["#"+id] = n;
    }
  }
  if (n == 0)
    noteholder.parentNode.removeChild(noteholder);
  else {
    // Process footnoterefs.
    for (i=0; i<spans.length; i++) {
      if (spans[i].className == "footnoteref") {
        var href = spans[i].getElementsByTagName("a")[0].getAttribute("href");
        href = href.match(/#.*/)[0];  // Because IE return full URL.
        n = refs[href];
        spans[i].innerHTML =
          "[<a href='#_footnote_" + n +
          "' title='View footnote' class='footnote'>" + n + "</a>]";
      }
    }
  }
},

install: function(toclevels) {
  var timerId;

  function reinstall() {
    asciidoc.footnotes();
    if (toclevels) {
      asciidoc.toc(toclevels);
    }
  }

  function reinstallAndRemoveTimer() {
    clearInterval(timerId);
    reinstall();
  }

  timerId = setInterval(reinstall, 500);
  if (document.addEventListener)
    document.addEventListener("DOMContentLoaded", reinstallAndRemoveTimer, false);
  else
    window.onload = reinstallAndRemoveTimer;
}

}
asciidoc.install();
/*]]>*/
</script>
</head>
<body class="article">
<div id="header">
<h1>RNA-Seq De novo Assembly Using Trinity</h1>
</div>
<div id="content">
<div id="preamble">
<div class="sectionbody">
<div class="paragraph"><p><span class="image">
<img src="images/TrinityCompositeLogo.png" alt="Trinity_logo" />
</span></p></div>
<div class="paragraph"><p>Trinity, developed at the <a href="http://www.broadinstitute.org">Broad Institute</a> and the <a href="http://www.cs.huji.ac.il">Hebrew University of Jerusalem</a>, represents a novel method for the efficient and robust de novo reconstruction of transcriptomes from RNA-seq data. Trinity combines three independent software modules: Inchworm, Chrysalis, and Butterfly, applied sequentially to process large volumes of RNA-seq reads. Trinity partitions the sequence data into many individual de Bruijn graphs, each representing the transcriptional complexity at at a given gene or locus, and then processes each graph independently to extract full-length splicing isoforms and to tease apart transcripts derived from paralogous genes.  Briefly, the process works like so:</p></div>
<div class="ulist"><ul>
<li>
<p>
<strong>Inchworm</strong> assembles the RNA-seq data into the unique sequences of transcripts, often generating full-length transcripts for a dominant isoform, but then reports just the unique portions of alternatively spliced transcripts.
</p>
</li>
<li>
<p>
<strong>Chrysalis</strong> clusters the Inchworm contigs into clusters and constructs complete de Bruijn graphs for each cluster.  Each cluster represents the full transcriptonal complexity for a given gene (or sets of genes that share sequences in common).  Chrysalis then partitions the full read set among these disjoint graphs.
</p>
</li>
<li>
<p>
<strong>Butterfly</strong> then processes the individual graphs in parallel, tracing the paths that reads and pairs of reads take within the graph, ultimately reporting full-length transcripts for alternatively spliced isoforms, and teasing apart transcripts that corresponds to paralogous genes.
</p>
</li>
</ul></div>
<div class="paragraph"><p>Trinity was published in <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3571712/">Nature Biotechnology</a>.  Our protocol for transcriptome assembly and downstream analysis is now published in <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3875132/">Nature Protocols</a></p></div>
<div class="paragraph"><p>The Trinity software package can be downloaded <a href="https://sourceforge.net/projects/trinityrnaseq/files/">here</a>.</p></div>
<div class="paragraph"><p><a href="http://www.broadinstitute.org/partnerships/education/broade/trinity-screencast">Screencast videos</a> are available to introduce you to Trinity and its various components. Also, hands-on tutorials for Trinity and Tuxedo are available as part of our <a href="rnaseq_workshop.html">RNA-Seq Workshop</a>.</p></div>
</div>
</div>
<div class="sect1">
<h2 id="_table_of_contents">Table of Contents</h2>
<div class="sectionbody">
<div class="ulist"><ul>
<li>
<p>
<a href="#installation">Installing Trinity</a>
</p>
</li>
<li>
<p>
<a href="#running_trinity">Running Trinity</a>
</p>
<div class="ulist"><ul>
<li>
<p>
<a href="#typical_usage">Typical Trinity Command Line</a>
</p>
</li>
<li>
<p>
<a href="#typical_options">Options to consider when running Trinity</a>
</p>
<div class="ulist"><ul>
<li>
<p>
<a href="#trimmomatic">Quality trimming using Trimmomatic</a>
</p>
</li>
<li>
<p>
<a href="#insilinorm">In silico Normalization of Reads (for assembly of (hundreds of millions to billions of reads)</a>
</p>
</li>
<li>
<p>
<a href="#jaccard_clip">Minimizing falsely fused transcripts from gene-dense genomes</a>
</p>
</li>
<li>
<p>
<a href="#genome_guided">Genome-guided Trinity</a>
</p>
</li>
<li>
<p>
<a href="#genome_annotation">Comprehensive transcriptome-based genome annotation using Trinity and PASA</a>
</p>
</li>
</ul></div>
</li>
<li>
<p>
<a href="#trinity_output">Output of Trinity</a>
</p>
</li>
<li>
<p>
<a href="#compute_requirements">Hardware and Configuration Requirements</a>
</p>
</li>
<li>
<p>
<a href="#monitoring_trinity">Monitoring the Progress of Trinity</a>
</p>
</li>
<li>
<p>
<a href="#sample_data">Running Trinity on Sample Data</a>
</p>
</li>
</ul></div>
</li>
<li>
<p>
<a href="#Downstream_analyses">Post-assembly Transcriptome Analysis</a>
</p>
<div class="ulist"><ul>
<li>
<p>
<a href="analysis/abundance_estimation.html">Abundance Estimation using RSEM or eXpress, and Visualization using IGV</a>
</p>
</li>
<li>
<p>
<a href="analysis/diff_expression_analysis.html">Differential Expression Analysis using Bioconductor</a>
</p>
</li>
<li>
<p>
<a href="analysis/extract_proteins_from_trinity_transcripts.html">Protein-coding Region Identification Using TransDecoder</a>
</p>
</li>
<li>
<p>
<a href="http://trinotate.sf.net">Functional Annotation Using Trinotate</a>
</p>
</li>
<li>
<p>
<a href="analysis/full_length_transcript_analysis.html">Full-length Transcript Analysis</a>
</p>
</li>
</ul></div>
</li>
<li>
<p>
<a href="advanced_trinity_guide.html">Advanced Guide to Trinity</a>
</p>
</li>
<li>
<p>
<a href="trinity_faq.html">Frequently Asked Questions</a>
</p>
</li>
<li>
<p>
<a href="#trinity_tidbits">Trinity Tidbits</a>
</p>
</li>
<li>
<p>
<a href="#trinity_developers">Trinity Developers Group</a>
</p>
</li>
<li>
<p>
<a href="#contact_us">Contact Us</a>
</p>
</li>
<li>
<p>
<a href="#referencing_trinity">Referencing Trinity</a>
</p>
</li>
</ul></div>
</div>
</div>
<div class="sect1">
<h2 id="installation">Installing Trinity</h2>
<div class="sectionbody">
<div class="sect2">
<h3 id="_local_installation_of_trinity_on_a_high_memory_linux_server">Local Installation of Trinity on a High-memory Linux Server</h3>
<div class="paragraph"><p>After <a href="https://sourceforge.net/projects/trinityrnaseq/files/">downloading</a> the sofware to a Linux server, simply type <em>make</em> in the base installation directory.  This should build Inchworm and Chrysalis, both written in C++.  Butterfly should not require any special compilation, as its written in Java and already provided as portable precompiled software.</p></div>
<div class="paragraph"><p>Trinity has been tested and is supported on Linux.</p></div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="Computing_Grid">Adapting Trinity to a computing grid for parallel processing of naively parallel steps</h2>
<div class="sectionbody">
<div class="paragraph"><p>Trinity has many parallel-components, all of which can benefit from having multiple CPUs on a single server, but there are also cases such as in Chrysalis and Butterfly where tens of thousands to hundreds of thousands of commands can be executed in parallel, each having independent inputs and outputs.  These naively-parallel commands can be most efficiently computed in the context of a compute farm, submitting each of the commands (or batches of them) to individual nodes on the computing grid.  There are several different computing grid job management systems that are in common use, such as SGE or LSF.</p></div>
<div class="paragraph"><p>Trinity currently supports both SGE and LSF.  To leverage either, simply run <em>Trinity --grid_conf your_conf_file.txt</em>, where your_conf_file.txt is a very simple configuration file that indicates parameters for the grid job submission. For example, at the Broad and using LSF, a configuration file might contain the following:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>#-------------------------------------------------------------------------------------------
# grid type:
grid=LSF</code></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><code># template for a grid submission
cmd=bsub -q regevlab -R "rusage[mem=10]"
# note -e error.file -o out.file are set internally, so dont set them in the above cmd.</code></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><code># uses the LSF feature to pre-exec and check that the file system is mounted before executing.
# this helps when you have some misbehaving grid nodes that lost certain file mounts.
mount_test=T</code></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><code>##########################################################################################
# settings below configure the Trinity job submission system, not tied to the grid itself.
##########################################################################################</code></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><code># number of grid submissions to be maintained at steady state by the Trinity submission system
max_nodes=500</code></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><code># number of commands that are batched into a single grid submission job.
cmds_per_node=100</code></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><code>#--------------------------------------------------------------------------------------------</code></pre>
</div></div>
<div class="paragraph"><p>where the above indicates that LSF is the grid type (either LSF or SGE are supported), the queue to submit to is our <em>regevlab</em> named queue, and memory is set to 10 gigabytes. Up to 500 jobs will be submitted at any given time (throttled by the Trinity-included job management system), and the jobs are batched at 10 commands per submission (so, for example, 10 butterfly jobs will be submitted as a single grid job, each being executed serially).</p></div>
<div class="paragraph"><p>For SGE, at the Broad Institute, we might specify a configuration:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>#--------------------------------------------------------------------------------------------
# grid type:
grid=SGE
# template for a grid submission
cmd=qsub -V -cwd
# number of grid submissions to be maintained at steady state by the Trinity submission system
max_nodes=500
# number of commands that are batched into a single grid submission job.
cmds_per_node=1
#--------------------------------------------------------------------------------------------</code></pre>
</div></div>
<div class="paragraph"><p>where, SGE is indicated as the grid type.  We don&#8217;t need to specify a queue name, apparently, as it gets submitted to the default queue, and the default memory allocation is sufficient. The project_code can also be left blank unless your SGE configuration requires it.  The maximum number of nodes to throttle the jobs at (500) and the number of commands executed in a single grid job (10) is the same as what we show above for our LSF configuration.</p></div>
<div class="paragraph"><p>Likewise, for SLURM, we have:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>#---------------------------------------------------------------------------------------------
# grid type:
grid=SLURM
# template for a grid submission
cmd=sbatch -p queue_name --mem=10000 --time=02:00:00
# number of grid submissions to be maintained at steady state by the Trinity submission system
max_nodes=4000
# number of commands that are batched into a single grid submission job.
cmds_per_node=20
#----------------------------------------------------------------------------------------------</code></pre>
</div></div>
<div class="paragraph"><p>Example configuration files are provided under $TRINITY_HOME/htc_conf</p></div>
<div class="paragraph"><p>Similar support for PBS is underway and planned for a future Trinity release.</p></div>
<div class="sect2">
<h3 id="RunElsewhere">Using a Freely Available Trinity Installation on High Performance Computing Systems</h3>
<div class="ulist"><ul>
<li>
<p>
Trinity is available on XSEDE&#8217;s Blacklight server at the <a href="http://www.psc.edu/">Pittsburgh Supercomputer Center</a>.  Information on how researchers in the USA can get a FREE account and to run Trinity on Blacklight (which has up to 16TB of RAM!) is provided <a href="http://trinity-use-on-blacklight-psc.wikispaces.com/Trinity+Usage+on+Blacklight">here</a>. Thanks to Phil Blood and Brian Cougar for maintaining this installation and making services available.
</p>
</li>
<li>
<p>
<a href="http://diagcomputing.org/">The Data Intensive Acadmeic Grid (DIAG)</a> provides <strong>FREE ACCESS TO ALL RESEARCHERS</strong> high memory servers and data storage for academic research. Trinity is supported as one of the pre-installed applications. The guide for running Trinity on DIAG is <a href="http://wiki.diagcomputing.org/index.php/Trinity">here</a>. Thanks to Anup Mahurkar and Joshua Orvis for support.
</p>
</li>
</ul></div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="running_trinity">Running Trinity</h2>
<div class="sectionbody">
<div class="paragraph"><p>Trinity is run via the script: <em>Trinity</em> found in the base installation directory.</p></div>
<div class="paragraph"><p>Usage info is as follows:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>###############################################################################
#
#     ______  ____   ____  ____   ____  ______  __ __
#    |      ||    \ |    ||    \ |    ||      ||  |  |
#    |      ||  D  ) |  | |  _  | |  | |      ||  |  |
#    |_|  |_||    /  |  | |  |  | |  | |_|  |_||  ~  |
#      |  |  |    \  |  | |  |  | |  |   |  |  |___, |
#      |  |  |  .  \ |  | |  |  | |  |   |  |  |     |
#      |__|  |__|\_||____||__|__||____|  |__|  |____/
#
###############################################################################
#
# Required:
#
#  --seqType &lt;string&gt;      :type of reads: ( fa, or fq )
#
#  --JM &lt;string&gt;            :(Jellyfish Memory) number of GB of system memory to use for
#                            k-mer counting by jellyfish  (eg. 10G) *include the 'G' char
#
#  If paired reads:
#      --left  &lt;string&gt;    :left reads, one or more (separated by space)
#      --right &lt;string&gt;    :right reads, one or more (separated by space)
#
#  Or, if unpaired reads:
#      --single &lt;string&gt;   :single reads, one or more (note, if single file contains pairs, can use flag: --run_as_paired )
#
####################################
##  Misc:  #########################
#
#  --SS_lib_type &lt;string&gt;          :Strand-specific RNA-Seq read orientation.
#                                   if paired: RF or FR,
#                                   if single: F or R.   (dUTP method = RF)
#                                   See web documentation.
#
#  --CPU &lt;int&gt;                     :number of CPUs to use, default: 2
#  --min_contig_length &lt;int&gt;       :minimum assembled contig length to report
#                                   (def=200)
#
#  --genome &lt;string&gt;               :genome guided mode, provide path to genome fasta file (see genome-guided param section)
#
#  --jaccard_clip                  :option, set if you have paired reads and
#                                   you expect high gene density with UTR
#                                   overlap (use FASTQ input file format
#                                   for reads).
#                                   (note: jaccard_clip is an expensive
#                                   operation, so avoid using it unless
#                                   necessary due to finding excessive fusion
#                                   transcripts w/o it.)
#
#  --trimmomatic                   :run Trimmomatic to quality trim reads
#                                        see '--quality_trimming_params' under full usage info for tailored settings.
#
#
#  --normalize_reads               :run in silico normalization of reads. Defaults to max. read coverage of 50.
#                                       see '--normalize_max_read_cov' under full usage info for tailored settings.
#
#
#  --output &lt;string&gt;               :name of directory for output (will be
#                                   created if it doesn't already exist)
#                                   default( your current working directory: "/Users/bhaas/SVN/trinityrnaseq/trunk/trinity_out_dir" )
#
#  --full_cleanup                  :only retain the Trinity fasta file, rename as ${output_dir}.Trinity.fasta
#
#  --cite                          :show the Trinity literature citation
#
#  --version                       :reports Trinity version (BLEEDING_EDGE) and exits.
#
#  --show_full_usage_info          :show the many many more options available for running Trinity (expert usage).
#
###############################################################################
#
#  *Note, a typical Trinity command might be:
#        Trinity --seqType fq --JM 100G --left reads_1.fq  --right reads_2.fq --CPU 6
#
#     see: /Users/bhaas/SVN/trinityrnaseq/trunk/sample_data/test_Trinity_Assembly/
#          for sample data and 'runMe.sh' for example Trinity execution
#     For more details, visit: http://trinityrnaseq.sf.net
#
###############################################################################</code></pre>
</div></div>
<div class="admonitionblock">
<table><tr>
<td class="icon">
<div class="title">Note</div>
</td>
<td class="content">Trinity performs best with strand-specific data, in which case sense and antisense transcripts can be resolved.  For protocols on strand-specific RNA-Seq, see: <a href="http://www.ncbi.nlm.nih.gov/pubmed/21943893">Borodina T, Adjaye J, Sultan M. A strand-specific library preparation protocol for RNA sequencing. Methods Enzymol. 2011;500:79-98. PubMed PMID: 21943893</a>.</td>
</tr></table>
</div>
<div class="paragraph"><p>If you have strand-specific data, specify the library type.  There are four library types:</p></div>
<div class="ulist"><ul>
<li>
<p>
Paired reads:
</p>
<div class="ulist"><ul>
<li>
<p>
<strong>RF</strong>: first read (/1) of fragment pair is sequenced as anti-sense (reverse(<strong>R</strong>)), and second read (/2) is in the sense strand (forward(<strong>F</strong>)); typical of the dUTP/UDG sequencing method.
</p>
</li>
<li>
<p>
<strong>FR</strong>: first read (/1) of fragment pair is sequenced as sense (forward), and second read (/2) is in the antisense strand (reverse)
</p>
</li>
</ul></div>
</li>
<li>
<p>
Unpaired (single) reads:
</p>
<div class="ulist"><ul>
<li>
<p>
<strong>F</strong>: the single read is in the sense (forward) orientation
</p>
</li>
<li>
<p>
<strong>R</strong>: the single read is in the antisense (reverse) orientation
</p>
</li>
</ul></div>
</li>
</ul></div>
<div class="paragraph"><p>By setting the <strong>--SS_lib_type</strong> parameter to one of the above, you are indicating that the reads are strand-specific.  By default, reads are treated as not strand-specific.</p></div>
<div class="paragraph"><p><span class="image">
<img src="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC3875132/bin/nihms-537313-f0006.jpg" alt="strand-specific library type" />
</span></p></div>
<div class="paragraph"><p>Other important considerations:</p></div>
<div class="ulist"><ul>
<li>
<p>
Whether you use Fastq or Fasta formatted input files, be sure to keep the reads oriented as they are reported by Illumina, if the data are strand-specific. This is because, Trinity will properly orient the sequences according to the specified library type.  If the data are not strand-specific, now worries because the reads will be parsed in both orientations.
</p>
</li>
<li>
<p>
If you have both paired and unpaired data, and the data are NOT strand-specific, you can combine the unpaired data with the left reads of the paired fragments.  Be sure that the unpaired reads have a /1 as a suffix to the accession value similarly to the left fragment reads.  The right fragment reads should all have /2 as the accession suffix.  Then, run Trinity using the --left and --right parameters as if all the data were paired.
</p>
</li>
<li>
<p>
If you have multiple paired-end library fragment sizes, set the <em>--group_pairs_distance</em> according to the larger insert library.  Pairings that exceed that distance will be treated as if they were unpaired by the Butterfly process.
</p>
</li>
<li>
<p>
by setting the <em>--CPU option</em>, you are indicating the maximum number of threads to be used by processes within Trinity. Note that Inchworm alone will be internally capped at 6 threads, since performance will not improve for this step beyond that setting)
</p>
</li>
</ul></div>
</div>
</div>
<div class="sect1">
<h2 id="typical_usage">Typical Trinity Command Line</h2>
<div class="sectionbody">
<div class="paragraph"><p>A typical Trinity command for assembling non-strand-specific RNA-seq data would be like so, running the entire process on a single high-memory server (aim for <sub>1G RAM per </sub>1M ~76 base Illumina paired reads, but often <strong>much</strong> less memory is required):</p></div>
<div class="paragraph"><p>Run Trinity like so:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>Trinity --seqType fq --JM 50G --left reads_1.fq  --right reads_2.fq --CPU 6</code></pre>
</div></div>
<div class="paragraph"><p>If you have multiple sets of fastq files, such as corresponding to multiple tissue types or conditions, etc., you can indicate them to Trinity like so:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>Trinity --seqType fq --JM 50G  --left condA_1.fq,condB_1.fq,condC_1.fq --right condA_2.fq,condB_2.fq,condC_2.fq --CPU 6</code></pre>
</div></div>
<div class="paragraph"><p>Also note that fastq files can be gzip-compressed, in which case they should require a <em>.gz</em> extension.</p></div>
<div class="paragraph"><p>Example data and sample pipeline are provided and described <a href="#sample_data">here</a>.</p></div>
</div>
</div>
<div class="sect1">
<h2 id="typical_options">Options to Consider when Running Trinity</h2>
<div class="sectionbody">
<div class="paragraph"><p>Trinity includes additional options to automate various aspects of RNA-Seq read processing that should be considered prior to executing the de novo assembly. This includes quality trimming of reads (using <a href="http://www.usadellab.org/cms/?page=trimmomatic">Trimmomatic</a>), or in silico normalization of the total reads to reduce the number of reads that are subject to de novo assembly, improving on assembly run-time.  Also, if transcripts are derived from a compact genome where overlapping UTRs are common, options are provided to mitigate the assembly of falsely end-to-end fused transcripts by analyzing the consistency of the read pairings across the length of the transcripts. These options are each detailed below.</p></div>
<div class="sect2">
<h3 id="trimmomatic">Quality trimming using Trimmomatic</h3>
<div class="paragraph"><p>To perform quality trimming of inputted fastq files, use <em>Trinity --trimmomatic</em>.  The default settings for quality trimming are described under the full usage info for Trinity (use <em>Trinity --show_full_usage_info</em> for complete usage info):</p></div>
<div class="literalblock">
<div class="content">
<pre><code>################################################################################
#### Quality Trimming Options ####
#
#  --quality_trimming_params &lt;string&gt;   defaults to: "LEADING:5 TRAILING:5 MINLEN:36"
#
################################################################################</code></pre>
</div></div>
<div class="paragraph"><p>The various options that are available for the Trimmomatic software are described on the <a href="http://www.usadellab.org/cms/?page=trimmomatic">Trimmomatic software website</a>.  The Trimmomatic software is bundled as a trinity plugin for convenience.</p></div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="insilinorm">Assembling Large RNA-Seq Data Sets (hundreds of millions to billions of reads)</h2>
<div class="sectionbody">
<div class="paragraph"><p>If you have especially large RNA-Seq data sets involving many hundreds of millions of reads to billions of reads, consider performing an in silico normalization of the full data set using <em>Trinity --normalize_reads</em>.  The default normalization process should work well for most data sets. If you prefer to manually set normalization-related parameters, you can find the options under the full Trinity usage info:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>################################################################################
####  In silico Read Normalization Options ###
#
#  --normalize_max_read_cov &lt;int&gt;       defaults to 50
#  --normalize_by_read_set              run normalization separate for each pair of fastq files,
#                                       then one final normalization that combines the individual normalized reads.
#                                       Consider using this if RAM limitations are a consideration.
#
################################################################################</code></pre>
</div></div>
<div class="paragraph"><p>If you are interested in running the normalization utility outside of Trinity, you can run it directly as described <a href="trinity_insilico_normalization.html">here</a>.</p></div>
<div class="sect2">
<h3 id="jaccard_clip">Minimizing Fusion Transcripts Derived from Gene Dense Genomes (using --jaccard_clip)</h3>
<div class="paragraph"><p>If your transcriptome RNA-seq data are derived from a gene-dense compact genome, such as from fungal genomes, where transcripts may often overlap in UTR regions, you can minimize fusion transcripts by leveraging the <strong>--jaccard_clip</strong> option if you have paired reads.  Trinity will examine the consistency of read pairings and fragment transcripts at positions that have little read-pairing support.  In expansive genomes of vertebrates and plants, this is unnecessary and not recommended.  In compact fungal genomes, it is highly recommended.  In addition to requiring paired reads, you must also have the <a href="http://bowtie-bio.sourceforge.net/index.shtml">Bowtie</a> short read aligner installed.  As part of this analysis, reads are aligned to the Inchworm contigs using Bowtie, and read pairings are examined across the Inchworm contigs, and contigs are clipped at positions of low pairing support.  These clipped Inchworm contigs are then fed into Chrysalis for downstream processing.</p></div>
<div class="paragraph"><p>Note, by using strand-specific RNA-Seq data alone, you should greatly mitigate the incorrect fusion of minimally overlapping transcripts.</p></div>
</div>
<div class="sect2">
<h3 id="genome_guided">Genome-guided Trinity</h3>
<div class="paragraph"><p>If a genome sequence is available, Trinity offers a method whereby reads are first aligned to the genome, partitioned according to locus, followed by de novo transcriptome assembly at each locus.</p></div>
<div class="paragraph"><p>Users can provide read alignments to Trinity as a coordinate-sorted bam file, or Trinity will run GSNAP directly to generate the read alignments (requires installation of <a href="http://research-pub.gene.com/gmap/">GSNAP</a>).</p></div>
<div class="paragraph"><p>To run Genome-guided Trinity and have Trinity execute GSNAP to align the reads, run Trinity like so:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>Trinity --genome genome.fasta --genome_guided_max_intron 10000 --genome_guided_sort_buffer 10G \
        --genome_guided_CPU 4 --GMAP_CPU 10 \
        --seqType fq --JM 2G --left reads_1.fq  --right reads_2.fq --CPU 10</code></pre>
</div></div>
<div class="paragraph"><p>Of course, use a maximum intron length that makes most sense given your targeted organism.</p></div>
<div class="paragraph"><p>Each of the sets of genome-partitioned reads will be executed locally at the --CPU jobs in parallel setting.  The --JM and --genome_guided_CPU parameters correspond to the RAM and thread allocations to each of these smaller assembly jobs. For example, if you&#8217;re running --CPU 10 parallel assemblies, with each allocating --JM 10G of RAM and using --genome_guided_CPU 4 number of threads. (10G x 10 = 100G of RAM, and 4 * 10 = 40 threads).  Use the --grid_conf parameter to execute the parallel jobs using SGE or LSF rather than executing them locally via multithreading.</p></div>
<div class="paragraph"><p>The --GMAP_CPU parameter allows you to separately configure the GMAP alignment process to use a different number of threads than the other parts of the process.  Note, the --GMAP_CPU and the --genome_guided_CPU parameters default to the ---CPU setting, if you choose not to specify them directly.</p></div>
<div class="paragraph"><p>The --genome_guided_sort_buffer setting is used during the initial stage of preparing the read paritions. It will default to the --JM setting. Because the --JM setting for the individual assembly commands can be relatively small compared to what is needed for a full de novo assembly, you may wish to keep the --JM value small (ie. 2G) while setting the --genome_guided_sort_buffer to a much higher value (ie. 10G or more).</p></div>
<div class="paragraph"><p>If you decide to run your own RNA-Seq alignments separately using your method of choice, simply provide that coordinate-sorted bam file to Trinity like so:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>Trinity --genome genome.fasta --genome_guided_use_bam rnaseq_alignments.csorted.bam \
        --genome_guided_max_intron 10000 --genome_guided_sort_buffer 10G \
        --genome_guided_CPU 4 \
        --seqType fq --JM 2G --left reads_1.fq  --right reads_2.fq --CPU 10</code></pre>
</div></div>
<div class="admonitionblock">
<table><tr>
<td class="icon">
<div class="title">Note</div>
</td>
<td class="content">Be sure to include additional options such as <em>--SS_lib_type</em> and <em>--jaccard_clip</em> where appropriate.  If quality trimming or normalization are indicated, these processes will be performed prior to aligning the reads to the genome.</td>
</tr></table>
</div>
<div class="paragraph"><p>In summary, the genome-guided Trinity involves two major phases.  The first phase involves partitioning genome-aligned reads into subsets that will each be targeted for independent Trinity de novo assembly.</p></div>
<div class="paragraph"><p>This first phase of either aligning reads (or using an existing coordinate-sorted bam file) happens on a single server, runs multithreaded, and leverages the parameters:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>--genome &lt;string&gt;  (ie. genome.fasta)
--genome_guided_max_intron &lt;int&gt;  (ie. '10000')
--genome_guided_sort_buffer &lt;string&gt; (ie. '10G')
--CPU &lt;int&gt;  (ie. 10)
--GMAP_CPU &lt;int&gt;  (ie. 10, defaults to --CPU setting)
   or use: --genome_guided_use_bam &lt;string&gt;   (ie. gsnap.coordSorted.bam)</code></pre>
</div></div>
<div class="paragraph"><p>The second phase involves running Trinity de novo assembly on each of the partitioned sets of reads.  If you end up with tens of thousands or hundreds of thousands of sets of partitioned reads, this means that you&#8217;ll have that large number of de novo assemblies to execute (in parallel). Each of these parallel-executed commands leverages the parameters:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>--genome_guided_CPU &lt;int&gt;  (ie. 4, *beware* that this defaults to --CPU)
--JM &lt;string&gt;   (ie. '2G', note that not much RAM is required for assembly of these relative small sets of reads)</code></pre>
</div></div>
<div class="paragraph"><p>If you specify --grid_conf &lt;string&gt;, then the commands in this second phase will be executed in parallel on your compute farm, using LSF, SGE, or other supported method.  Otherwise, these commands will be executed locally using our Parafly parallel command processor, throttled at --CPU number of parallel processes.</p></div>
</div>
<div class="sect2">
<h3 id="genome_annotation">Comprehensive transcriptome-based genome annotation using Trinity and PASA</h3>
<div class="paragraph"><p>The Trinity-reconstructed transcripts can be used to annotate genomes using PASA.  Documentation for this is provided on the PASA website under <a href="http://pasa.sourceforge.net/#A_ComprehensiveTranscriptome">Build a Comprehensive Transcriptome Database Using Genome-guided and De novo RNA-Seq Assembly</a> link.</p></div>
</div>
</div>
</div>
<div class="sect1">
<h2 id="trinity_output">Output of Trinity</h2>
<div class="sectionbody">
<div class="paragraph"><p>When Trinity completes, it will create a <em>Trinity.fasta</em> output file in the <em>trinity_out_dir/</em> output directory (or output directory you specify).</p></div>
<div class="paragraph"><p>Trinity groups transcripts into clusters based on shared sequence content. Such a transcript cluster is very loosely referred to as a <em>gene</em>. This information is encoded in the Trinity fasta accession.  An example Fasta entry for one of the transcripts is formatted like so:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>&gt;c115_g5_i1 len=247 path=[31015:0-148 23018:149-246]
AATCTTTTTTGGTATTGGCAGTACTGTGCTCTGGGTAGTGATTAGGGCAAAAGAAGACAC
ACAATAAAGAACCAGGTGTTAGACGTCAGCAAGTCAAGGCCTTGGTTCTCAGCAGACAGA
AGACAGCCCTTCTCAATCCTCATCCCTTCCCTGAACAGACATGTCTTCTGCAAGCTTCTC
CAAGTCAGTTGTTCACAGGAACATCATCAGAATAAATTTGAAATTATGATTAGTATCTGA
TAAAGCA</code></pre>
</div></div>
<div class="paragraph"><p>The accession encodes the Trinity <em>gene</em> and <em>isoform</em> information. In the example above, the accession <em>c115_g5_i1</em> indicates Trinity read cluster <em>c115</em>, gene <em>g5</em>, and isoform <em>i1</em>.  Because a given run of trinity involves many many clusters of reads, each of which are assembled separately, and because the <em>gene</em> numberings are unique within a given processed read cluster, the <em>gene</em> identifier should be considered an aggregate of the read cluster and corresponding gene identifier, which in this case would be <em>c115_g5</em>.</p></div>
<div class="paragraph"><p>So, in summary, the above example corresponds to <em>gene id: c115_g5</em> encoding <em>isoform id: c115_g5_i1</em>.</p></div>
<div class="paragraph"><p>Obtain basic stats for the number of <em>genes</em> and <em>isoforms</em> and contiguity of the assembly by running:</p></div>
<div class="literalblock">
<div class="content">
<pre><code>% $TRINITY_HOME/util/TrinityStats.pl trinity_out_dir/Trinity.fasta</code></pre>
</div></div>
<div class="paragraph"><p>with output (example from assembling our 10M Schizosaccharoymyces pombe data set):</p></div>
<div class="literalblock">
<div class="content">
<pre><code>################################
## Counts of transcripts, etc.
################################
Total trinity 'genes':  8645
Total trinity transcripts:  9398
Percent GC: 37.59</code></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><code>########################################
Stats based on ALL transcript contigs:
########################################</code></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><code>Contig N10: 3838
Contig N20: 3124
Contig N30: 2629
Contig N40: 2243
Contig N50: 1936</code></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><code>Median contig length: 984
Average contig: 1251.23
Total assembled bases: 11759032</code></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><code>#####################################################
## Stats based on ONLY LONGEST ISOFORM per 'GENE':
#####################################################</code></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><code>Contig N10: 3848
Contig N20: 3124
Contig N30: 2630
Contig N40: 2250
Contig N50: 1937</code></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><code>Median contig length: 942
Average contig: 1227.97
Total assembled bases: 10615785</code></pre>
</div></div>
</div>
</div>
<div class="sect1">
<h2 id="compute_requirements">Hardware and Configuration Requirements</h2>
<div class="sectionbody">
<div class="paragraph"><p>The Inchworm and Chrysalis steps can be memory intensive.  A basic recommendation is to have <sub>1G of RAM per </sub>1M pairs of Illumina reads. Simpler transcriptomes (lower eukaryotes) require less memory than more complex transcriptomes such as from vertebrates.</p></div>
<div class="paragraph"><p>If you are able to run the entire Trinity process on a single high-memory multi-core server, indicate the number of butterfly processes to run in parallel by the --CPU parameter.</p></div>
<div class="paragraph"><p>Our experience is that the entire process can require ~1/2 hour to one hour per million pairs of reads in the current implementation (see <a href="trinity_faq.html">FAQ</a>).  We&#8217;re striving to improve upon both memory and time requirements.</p></div>
<div class="paragraph"><p>If you do not have direct access to a high memory machine (typically having 256G or 512G of RAM), consider <a href="#RunElsewhere">running Trinity on one of the externally available resources</a>.</p></div>
<div class="paragraph"><p>If you are limited to the amount of time available for executing Trinity (due to artificially imposed limits on a shared computing resource), you can aim to run Trinity in separate stages, where subsequent stages resume from the previous ones.  To do so, include the following options for each of the stages:</p></div>
<div class="ulist"><ul>
<li>
<p>
Stage 1: generate the kmer-catalog using jellyfish <em>--no_run_inchworm</em>
</p>
</li>
<li>
<p>
Stage 2: run Inchworm to generate draft contigs:  <em>--no_run_chrysalis</em>
</p>
</li>
<li>
<p>
Stage 2: Chrysalis clustering of inchworm contigs and mapping reads: <em>--no_run_quantifygraph</em>
</p>
</li>
<li>
<p>
Stage 3: Chrysalis deBruijn graph construction: <em>--no_run_butterfly</em>
</p>
</li>
<li>
<p>
Stage 4: Run butterfly, generate final Trinity.fasta file.  (exclude <em>--no_</em> options)
</p>
</li>
</ul></div>
</div>
</div>
<div class="sect1">
<h2 id="monitoring_trinity">Monitoring the Progress of Trinity</h2>
<div class="sectionbody">
<div class="paragraph"><p>Since Trinity can easily take several days to complete, it is useful to be able to monitor the process and to know at which stage (Inchworm, Chrysalis, Butterfly) Trinity is currently at.  There are a few general ways to do this:</p></div>
<div class="ulist"><ul>
<li>
<p>
by running <em>top</em>, you&#8217;ll be able to see which Trinity process is running and how much memory is being consumed.
</p>
</li>
<li>
<p>
other downstream process will generate standard output.  Be sure to capture <em>stdout</em> and <em>stderr</em> when you run the Trinity script.  The format for capturing both stdout and stderr depends on your SHELL.  Figure out what shell you have by running:
</p>
<div class="literalblock">
<div class="content">
<pre><code>env | grep SHELL</code></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><code>Using tcsh:</code></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><code>Trinity ... opts ... &gt; &amp; run.log &amp;</code></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><code>Using bash:</code></pre>
</div></div>
<div class="literalblock">
<div class="content">
<pre><code>Trinity ... opts ... &gt; run.log 2&gt;&amp;1 &amp;</code></pre>
</div></div>
</li>
</ul></div>
<div class="paragraph"><p>Note, under bash, to prevent the background process from being terminated once you close the shell, type <em>exit</em> to leave the shell, or explore alternatives such as <a href="http://www.serverwatch.com/tutorials/article.php/3935306/Detach-Processes-With-Disown-and-Nohup.htm">nohup, disown, or screen</a>.</p></div>
<div class="paragraph"><p>You can then <em>tail -f run.log</em> to follow the progress of the Trinity throughout the various stages.</p></div>
</div>
</div>
<div class="sect1">
<h2 id="sample_data">Running Trinity on Sample Data</h2>
<div class="sectionbody">
<div class="paragraph"><p>The Trinity software distribution includes sample data in the <em>sample_data/test_Trinity_Assembly/</em> directory. Simply run the included <em>runMe.sh</em> shell script to execute the Trinity assembly process with provided paired strand-specific Illumina data derived from mouse.  Running Trinity on the sample data requires &lt;~2G of RAM and should run on an ordinary desktop/laptop computer.  Run as <em>runMe.sh 1</em> to execute downstream analysis steps, including bowtie read alignment and RSEM-based abundance estimation, as described below.</p></div>
</div>
</div>
<div class="sect1">
<h2 id="Downstream_analyses">Downstream Analyses</h2>
<div class="sectionbody">
<div class="paragraph"><p>The following downstream analyses are supported as part of Trinity:</p></div>
<div class="ulist"><ul>
<li>
<p>
<a href="analysis/abundance_estimation.html">Abundance estimation using RSEM or eXpress, and visualization using IGV</a>.
</p>
</li>
<li>
<p>
<a href="analysis/diff_expression_analysis.html">Using EdgeR and Bioconductor for analyzing differentially expressed transcripts</a>.
</p>
</li>
<li>
<p>
<a href="analysis/extract_proteins_from_trinity_transcripts.html">Extract likely protein-coding regions from Trinity transcripts using TransDecoder</a>.
</p>
</li>
<li>
<p>
<a href="http://trinotate.sf.net">Functionally annotate transcripts and coding regions with Trinotate</a>.
</p>
</li>
<li>
<p>
<a href="analysis/full_length_transcript_analysis.html">Full-length transcript analysis for model and non-model transcriptomes</a>
</p>
</li>
</ul></div>
</div>
</div>
<div class="sect1">
<h2 id="advanced_guide">Want to know more?</h2>
<div class="sectionbody">
<div class="paragraph"><p>Visit the <a href="advanced_trinity_guide.html">Advanced Guide to Trinity</a> for more information regarding Trinity behavior, intermediate data files, and file formats.</p></div>
</div>
</div>
<div class="sect1">
<h2 id="faq">Frequently Asked Questions</h2>
<div class="sectionbody">
<div class="paragraph"><p>Visit the <a href="trinity_faq.html">Trinity FAQ</a> page.</p></div>
</div>
</div>
<div class="sect1">
<h2 id="trinity_tidbits">Trinity Tidbits</h2>
<div class="sectionbody">
<div class="ulist"><ul>
<li>
<p>
Trinity made the cover of the <a href="http://www.nature.com/nbt/journal/v29/n7/index.html">July 2011 NBT issue</a>. The Broad Institute&#8217;s <a href="http://www.broadinstitute.org/blog/suite-tools-takes-flight">blog</a> has a story on how the Trinity project came together. Nir Friedman, one of the project PIs, has a <a href="http://nirfriedmanlab.blogspot.com/2011/07/behind-cover.html">blog entry</a> describing the developmental process underlying the NBT cover design.
</p>
</li>
<li>
<p>
Trinity was shown to be the leading de novo transcriptome assembly tool as part of the <a href="http://www.the-dream-project.org/challanges/dream6-alternative-splicing-challenge">DREAM6 Alt-Splicing Challenge 2011</a>. Results were posted <a href="http://www.the-dream-project.org/result/alternative-splicing">here</a>.
</p>
</li>
<li>
<p>
<a href="http://scholar.google.com/scholar?oi=bibs&amp;hl=en&amp;cites=14735674943942667509">Google Scholar</a> shows how Trinity is being used by the community.
</p>
</li>
</ul></div>
</div>
</div>
<div class="sect1">
<h2 id="trinity_developers">Trinity Development Group</h2>
<div class="sectionbody">
<div class="paragraph"><p>Trinity is currently being maintained as an open source software project, primarily by the following contributors:</p></div>
<div class="ulist"><ul>
<li>
<p>
Josh Bowden, CSIRO
</p>
</li>
<li>
<p>
Brian Couger, Oklahoma State University
</p>
</li>
<li>
<p>
David Eccles, Max Planck Institute for Molecular Biomedicine, Mnster
</p>
</li>
<li>
<p>
Nir Friedman, Hebrew University (PI)
</p>
</li>
<li>
<p>
Manfred Grabherr, Biomedical Centre in Uppsala, Broad Institute
</p>
</li>
<li>
<p>
Brian Haas, Broad Institute
</p>
</li>
<li>
<p>
Robert Henschel, Indiana University
</p>
</li>
<li>
<p>
Jean-Marc Lassance, Harvard University
</p>
</li>
<li>
<p>
Matthias Lieber, Technische Universitat Dresden
</p>
</li>
<li>
<p>
Matthew MacManes, Berkeley
</p>
</li>
<li>
<p>
Joshua Orvis, Institute for Genome Sciences, Broad Institute
</p>
</li>
<li>
<p>
Michael Ott, CSIRO
</p>
</li>
<li>
<p>
Alexie Papanicolaou, CSIRO
</p>
</li>
<li>
<p>
Nathalie Pochet, Broad Institute
</p>
</li>
<li>
<p>
Aviv Regev, Broad Institute (PI)
</p>
</li>
<li>
<p>
Moran Yassour, Hebrew University, Broad Institute
</p>
</li>
<li>
<p>
Nathan Weeks, USDA-ARS
</p>
</li>
<li>
<p>
Rick Westerman, Purdue University
</p>
</li>
</ul></div>
<div class="paragraph"><p>Also, many valuable contributions come from the very active Trinity community via our mailing list (see below).</p></div>
</div>
</div>
<div class="sect1">
<h2 id="contact_us">Contact Us</h2>
<div class="sectionbody">
<div class="paragraph"><p>Questions, suggestions, comments, etc?</p></div>
<div class="paragraph"><p>Send email to <a href="https://sourceforge.net/mailarchive/forum.php?forum_name=trinityrnaseq-users">trinityrnaseq-users@lists.sf.net</a>.</p></div>
<div class="paragraph"><p>Subscribe to the email list <a href="https://lists.sourceforge.net/lists/listinfo/trinityrnaseq-users">here</a>.</p></div>
</div>
</div>
<div class="sect1">
<h2 id="referencing_trinity">Referencing Trinity</h2>
<div class="sectionbody">
<div class="paragraph"><p>Trinity can be referenced as:</p></div>
<div class="ulist"><ul>
<li>
<p>
Grabherr MG, Haas BJ, Yassour M, Levin JZ, Thompson DA, Amit I, Adiconis X, Fan L, Raychowdhury R, Zeng Q, Chen Z, Mauceli E, Hacohen N, Gnirke A, Rhind N,
di Palma F, Birren BW, Nusbaum C, Lindblad-Toh K, Friedman N, Regev A.
Full-length transcriptome assembly from RNA-seq data without a reference genome.
<a href="http://www.nature.com/nbt/journal/vaop/ncurrent/abs/nbt.1883.html">Nat Biotechnol. 2011 May 15;29(7):644-52</a>. doi: 10.1038/nbt.1883.
<a href="http://www.ncbi.nlm.nih.gov/pubmed/21572440">PubMed PMID: 21572440</a>.
</p>
</li>
</ul></div>
<div class="paragraph"><p>Protocol for using Trinity for de novo transcriptome assembly and downstream analyses:</p></div>
<div class="ulist"><ul>
<li>
<p>
Haas BJ, Papanicolaou A, Yassour M, Grabherr M, Blood PD, Bowden J, Couger MB,
Eccles D, Li B, Lieber M, Macmanes MD, Ott M, Orvis J, Pochet N, Strozzi F, Weeks
N, Westerman R, William T, Dewey CN, Henschel R, Leduc RD, Friedman N, Regev A.
De novo transcript sequence reconstruction from RNA-seq using the Trinity
platform for reference generation and analysis. <a href="http://www.nature.com/nprot/journal/v8/n8/full/nprot.2013.084.html">Nat Protoc. 2013 Aug;8(8):1494-512.</a> doi: 10.1038/nprot.2013.084. Epub 2013 Jul 11. PubMed PMID:
23845962.
</p>
</li>
</ul></div>
<div class="paragraph"><p>Performance tuning of Trinity is described in:</p></div>
<div class="ulist"><ul>
<li>
<p>
Henschel R, Lieber M, Wu L, Nista, PM, Haas BJ, LeDuc R.  Trinity RNA-Seq assembler performance optimization. XSEDE 2012 Proceedings of the 1st Conference of the Extreme Science and Engineering Discovery Environment: Bridging from the eXtreme to the campus and beyond. <a href="http://dx.doi.org/10.1145/2335755.2335842">ISBN: 978-1-4503-1602-6 doi&gt;10.1145/2335755.2335842</a>.
</p>
</li>
</ul></div>
<div class="paragraph"><p>A full list of references including Trinity, RSEM, and additional tools leveraged by Trinity can be obtained by running <em>Trinity --cite</em>.</p></div>
</div>
</div>
</div>
<div id="footnotes"><hr /></div>
<div id="footer">
<div id="footer-text">
Last updated 2014-10-02 18:53:42 EDT
</div>
</div>
</body>
</html>
